consensus/types/src/light_client_header.rs

```diff
@@ -1,26 +1,208 @@
-use crate::test_utils::TestRandom;
+use crate::beacon_block_body::NUM_BEACON_BLOCK_BODY_HASH_TREE_ROOT_LEAVES;
 use crate::BeaconBlockHeader;
+use crate::ForkName;
+use crate::ForkVersionDeserialize;
+use crate::{light_client_update::*, BeaconBlockBody};
+use crate::{
+    test_utils::TestRandom, EthSpec, ExecutionPayloadHeaderCapella, ExecutionPayloadHeaderDeneb,
+    FixedVector, Hash256, SignedBeaconBlock,
+};
+use merkle_proof::verify_merkle_proof;
 use serde::{Deserialize, Serialize};
+use serde_json;
 use ssz_derive::{Decode, Encode};
+use std::marker::PhantomData;
+use superstruct::superstruct;
 use test_random_derive::TestRandom;
+use tree_hash::TreeHash;
 
-#[derive(
-    Debug,
-    Clone,
-    PartialEq,
-    Serialize,
-    Deserialize,
-    Encode,
-    Decode,
-    TestRandom,
-    arbitrary::Arbitrary,
+#[superstruct(
+    variants(Altair, Capella, Deneb),
+    variant_attributes(
+        derive(
+            Default,
+            Debug,
+            Clone,
+            PartialEq,
+            Serialize,
+            Deserialize,
+            Encode,
+            Decode,
+            TestRandom,
+            arbitrary::Arbitrary,
+        ),
+        serde(bound = "E: EthSpec", deny_unknown_fields),
+        arbitrary(bound = "E: EthSpec"),
+    )
 )]
-pub struct LightClientHeader {
+#[derive(Debug, Clone, Serialize, Deserialize, Encode, Decode, arbitrary::Arbitrary, PartialEq)]
+#[serde(untagged)]
+#[serde(bound = "E: EthSpec", deny_unknown_fields)]
+#[ssz(enum_behaviour = "union")]
+#[arbitrary(bound = "E: EthSpec")]
+pub struct LightClientHeader<E: EthSpec> {
     pub beacon: BeaconBlockHeader,
+    #[superstruct(only(Capella, Deneb))]
+    pub execution_branch: FixedVector<Hash256, ExecutionPayloadProofLen>,
+    #[superstruct(
+        only(Capella),
+        partial_getter(rename = "execution_payload_header_capella")
+    )]
+    pub execution: ExecutionPayloadHeaderCapella<E>,
+    #[superstruct(only(Deneb), partial_getter(rename = "execution_payload_header_deneb"))]
+    pub execution: ExecutionPayloadHeaderDeneb<E>,
+
+    #[ssz(skip_serializing, skip_deserializing)]
+    pub phantom_data: PhantomData<E>,
+}
+
+impl<E: EthSpec> LightClientHeaderAltair<E> {
+    pub fn block_to_light_client_header(block: &SignedBeaconBlock<E>) -> Result<Self, Error> {
+        Ok(LightClientHeaderAltair {
+            beacon: block.message().block_header(),
+            phantom_data: PhantomData,
+        })
+    }
+
+    #[allow(dead_code)]
+    fn get_lc_execution_root(&self) -> Option<Hash256> {
+        None
+    }
+
+    #[allow(dead_code)]
+    fn is_valid_light_client_header(&self) -> Result<bool, Error> {
+        Ok(true)
+    }
```

Comment:
Looks like these functions aren't required by full nodes and can be excluded?
Was the idea to include it for the users of the `consensus-types` crate? This might be outside the responsibility of Lighthouse though and not sure if we should maintain something we don't use. They're also private so not accessible anyway?

Reply:
Yeah I added these because they were included in the spec. Wasn't really sure what our policy was here, but I can go ahead and delete